stm32: Improved CAN FD support.

iabdalkader opened the pull request:
### This PR does the following:
* Add support for CAN FD frames.
* Add support for Bitrate Switching (BRS) for data phase.
* Add support for extended (29-bit) frame identifiers.
* Add support for a bus with mixed Classic/FD nodes.
* Return the matching filter ID and type of received messages.
* Fix the maximum range for the nominal and data bit timing parameters.
* Optimize the Message RAM usage per FDCAN instance.
* Document the usage and different sections of the Message RAM.
* Update CAN docs.
* Fixes  #8169
* Fixes  #5621
* Fixes #3916
* Fixes openmv/openmv#1071

### Example usage:
* To send a classical CAN frame use the default send:
```Python
can.send('a' * 8, 1)
```

* To send an FD frame (long frame, up to 64 bytes) set `fdf` true:
```Python
can.send('a' * 64, 1, fdf=True)
```

* To use BRS (bitrate switching) set `brs_baudrate` (or brs timings), and set `brs` true, example:
```Python
# FD frame + BRS mode. 500 Kbit/s for arbitration phase, 1Mbit/s for data phase.
can = CAN(2, CAN.NORMAL, baudrate=500_000, brs_baudrate=1_000_000, sample_point=80)
can.send('a'*64, 1, fdf=True, brs=True)
```

* To use extended IDs set `extframe` true, example:
```Python
can = CAN(2, CAN.NORMAL, baudrate=500_000, sample_point=75)
# Note you should setfilter() on the receiver with extframe=True
can.send('HelloWorld', 0xFFFF, extframe=True)
```

* All options:
```Python
# FD frame + BRS mode + Extended frame ID. 500 Kbit/s for arbitration phase, 1Mbit/s for data phase.
can = CAN(2, CAN.NORMAL, baudrate=500_000, brs_baudrate=1_000_000, sample_point=80)
# Note you should setfilter() on the receiver with extframe=True
can.send('a'*64, 0xFFFF, fdf=True, brs=True, extframe=True)
```

### Testing:
I tested the following scenarios:
* Sending/Receiving classic CAN frames from FDCAN <-> FDCAN (STM32H7).
* Sending/Receiving FD (long) CAN frames (plus BRS) from FDCAN <-> FDCAN (STM32H7).
* Sending/Receiving classic CAN frame FDCAN <-> CAN (H7 <-> F7)
* Sending/Filtering&Receiving extended frame IDs (H7 <-> H7).

### Notes:

- This change doesn't break backwards compatibility with F7 CAN.
- ~I renamed `extframe` to `extid` because it never really enabled extended frames before, and classic CAN does Not support extended frames anyway.~
- After looking at the [CAN specs ](http://esd.cs.ucr.edu/webres/can20.pdf)and [CAN FD specs](https://can-newsletter.org/assets/files/ttmedia/raw/e5740b7b5781b8960f55efcc2b93edf8.pdf) I realized the specs refer to the extended (29-bit) ID as `extended frame` and `extended format` respectively. I still think `extid` is much better, and it is used in the HAL's CAN message structs, that said I reverted it.
- The `pyb_can_initfilterbanks()` functions is problematic, to say the least.  This function should Not be a class method for FDCAN because it can't be called before calling `can.init()` first to initialize the number of standard and extended filters and their Message RAM addresses/offsets based on the Message RAM configuration, and set which type of filter is used (although I init both standard and extended filters but we can optimize this). As for classic CAN, this function calls `HAL_CAN_ConfigFilter(NULL, &filter);` from `can.c`. If `CAN3` is defined, `HAL_CAN_ConfigFilter` will dereference a NULL CAN handle pointer.
- Following comments on #3916 in particular this comment https://github.com/micropython/micropython/issues/3916#issuecomment-405143240, I added `extframe` keyword argument to `send()`, `setfilter()` and `clearfilter()`, and removed it from `init()` because it's no longer needed. Also following comments on the same issue, `recv()` now returns a tuple for the matching filter ID, indicating the frame type (true for extended).

iabdalkader committed:
stm32: Enable CAN FD frame support and BRS.

* Enable CAN FD frame support and BRS.
* Optimize the message RAM usage per FDCAN instance.
* Document the usage and different sections of the Message RAM.

kwagyeman commented:
Can you fix extended frame mode too?

https://github.com/openmv/openmv/issues/1071

iabdalkader committed:
stm32: Add support for CAN FD extended frame ID.

JohnieBraaf commented:
No, #5508 remains open, but I guess it would be better to fix that in a seperate PR.
Thnx for pointing me to #8057, I was not aware of that PR, it addresses the same issue and more.

iabdalkader commented:
> No, #5508 remains open

Can you tell me how to reproduce this issue ? I can test with of H7, F7 and F4.

JohnieBraaf commented:
You can reproduce the issue with the following on both sender and receiver:

```
def receive(bus, reason):
    if reason == 0:
        print('message pending')

can1 = CAN(1, CAN.NORMAL, baudrate=1_000_000, sample_point=75, extframe=False, auto_restart=True)
can1.setfilter(bank=0, mode=pyb.CAN.MASK32, fifo=0, params=(0x0, 0x0))
can1.rxcallback(0, receive)
```

And test on both sides with:
```
can1.send(b'\xA0\x48\x08\x01\xB3\x11', 0x200)
```

What you will see is that on non-fd can interfaces the non extended id messages will never be received.
You can switch the exact same script to use extended ids (on both sides) and then the messages will be received.
`extid=True`

In fact the mesasges are received but they are filtered out, although the parameters 0x0, 0x0 should allow for them.
I have noticed that if the following lines are changed in pyb_can.c, the messages can be received withouth issues.

https://github.com/iabdalkader/micropython/blob/5d1ea1876f53b69c65431931405da50e045c2343/ports/stm32/pyb_can.c#L866-L869
```
        filter.FilterIdHigh = (mp_obj_get_int(params[0]) & 0x1FFFE000) >> 13;
        filter.FilterIdLow = (((mp_obj_get_int(params[0]) & 0x00001FFF) << 3) | 4) | rtr_masks[0];
        filter.FilterMaskIdHigh = (mp_obj_get_int(params[1]) & 0x1FFFE000) >> 13;
        filter.FilterMaskIdLow = (((mp_obj_get_int(params[1]) & 0x00001FFF) << 3) | 4) | rtr_masks[1];
```
changed to:
```
        filter.FilterIdHigh = 0x0000;
        filter.FilterIdLow = 0x0000;
        filter.FilterMaskIdHigh = 0x0000;
        filter.FilterMaskIdLow = 0x0000;
```

It seems to me that the masking is not working as it should, or perhaps should be different for non-extended ids.

JohnieBraaf commented:
While writing down the steps to reproduce and looking at the code I realized, I should have used MASK16 instead of MASK32.
Sorry for the confusion. I managed to receive non-extended id messages on non fd interfaces with the following filter:
`can1.setfilter(bank=0, mode=pyb.CAN.MASK16, fifo=0, params=(0x0, 0x0, 0x0, 0x0))`
Thanks for reaching out, the good news is, I can close the old issue.

iabdalkader commented:
That's great! Thanks for the feedback on the PR.

dpgeorge commented:
Can you please rebase this on latest master to pick up the other merged FDCAN PR?

iabdalkader commented:
> Can you please rebase this on latest master to pick up the other merged FDCAN PR?

Yes, I'll see what I can do, but I can't fix anything related to STM32G4. That said I documented the message ram usage very well and someone else would be able to fix it.

iabdalkader commented:
@dpgeorge rebased. I don't like setting the second set of bit timings (the BRS timings) in the `Init` struct in `pyb_can_init_helper()` before calling `can_init()`, but `can_init()` prototype will have to be changed for both CAN and FDCAN to accept more args (not used by `can.c`). Do you have any suggestions for a better way to do this ?

iabdalkader commented:
After looking at the [CAN specs ](http://esd.cs.ucr.edu/webres/can20.pdf)and [CAN FD specs](https://can-newsletter.org/assets/files/ttmedia/raw/e5740b7b5781b8960f55efcc2b93edf8.pdf) I realized the specs refer to the extended (29-bit) ID as `extended frame` and `extended format` respectively. I still think `extid` is much better, and it is used in the HAL's CAN message structs, that said I could reverted it.

iabdalkader committed:
stm32/pyb_can.c: Add support for bus with mixed Classic/FD nodes.

* A CAN bus can have mixed classic/FD nodes. Currently, CAN is
configured for either standard or extended ID.
* This patch allows extended IDs to be filtered and enabled on
a per message basis, in send() and set/clearfilter.

iabdalkader committed:
stm32/pyb_can.c: Return the filter ID and type of received messages.

iabdalkader commented:
Please see the updated notes and comments in the first message.

iabdalkader committed:
stm32/pyb_can.c: Define the maximum bit timing parameters.

* Define the maximum parameters for CAN/FDCAN nominal bit timing,
and for FDCAN data bit timing, for bit timing calculations.

iabdalkader commented:
@dpgeorge I've collected almost all? of the open CAN issues in this PR. All the linked issues should be fixed by this PR. I'm now working on the docs, the state of the CAN docs is really bad. Can you review & approve this so I can account for any API changes while I'm at it ?

iabdalkader committed:
docs/library/pyb.CAN.rst: Update CAN docs.

iabdalkader commented:
@dpgeorge docs updated.

dpgeorge commented:
Thanks for updating.  I'll take a look at this next.

dpgeorge commented:
After an initial review of this PR, here are my comments:
- It looks like it'll be too hard and confusing to retain backwards compatibility, so let's forget about that for the moment.
- It makes sense to support `extframe` as an argument to `send()` and the filter config functions/methods, that's a good change.  And hence also remove it from the constructor/`init()` method.
- While I agree that `extid` is more logical than `extframe`, the CAN specs and other internet resources on CAN are pretty clear about it being an "extended frame".  Even CAN FD calls it an FEFF (FD extended frame format).  So I think it's best to match the spec and continue to use `extframe`.  Just need to make the docs clear what this option does.
- I see the problem with `initfilterbanks()` being a static method.  But logically it should remain static because its configuring things before the CAN instances are created.  Maybe the way it can be implemented is to simply store the `nr` argument passed in and subsequently use that value to configure the CAN when the objects are actually instantiated.
- For `recv()`: to avoid allocating the 2-tuple, and making the return value nested tuples, I think the second element of the existing returned 4-tuple can become a bitmask with flags.  Currently it's a boolean for RTR, but it can be a bit-wise or of `RTR|EXTID`.  Alternatively make the 4-tuple a 5-tuple.

iabdalkader commented:
>  Just need to make the docs clear what this option does.

The docs are clear about it and were updated to reflect the changes (removed the `extframe` arg from `init()`, added it to `send()` and the other filter functions).

> For recv(): to avoid allocating the 2-tuple, and making the return value nested tuples, I think the second element of the existing returned 4-tuple can become a bitmask with flags. Currently it's a boolean for RTR, but it can be a bit-wise or of RTR|EXTID. Alternatively make the 4-tuple a 5-tuple.

I went with the 5-tuple option, docs/ code comments were both updated to reflect the changes.

>  Maybe the way it can be implemented is to simply store the nr argument passed in and subsequently use that value to configure the CAN when the objects are actually instantiated.

This is possible, however it's not that simple. The main issue here is that the CAN FD RAM (which is shared between all sections including filtering) is currently configured/split statically (i.e. the number of filters, TX/RX FIFO sizes, etc.. per instance is set statically in `fdcan.c`). Changing the number of std/ext filters dynamically, would require adjusting all the other sections according to the remaining memory, however, how do we determine which section gets increased/reduced based on the remaining available memory ? I think this will lead us to requiring that users provide the entire memory config/split in `initfilterbanks` before calling `init()`. Alternatively, an easier option is to just add an arg to enable using the entire message RAM for 1 CAN instance (use the same memory offset for CAN1/2), which would give the user double the number of filters, double TX/RX FIFO sizes and so on.

iabdalkader commented:
ping @dpgeorge

lorismich commented:
Please 🙏 @dpgeorge

dpgeorge commented:
It looks like the only outstanding item here is what to do with `initfilterbanks`.  Based in the discussion above, I have 2 suggestions for what to do.

#### Option 1
- rename `initfilterbanks` to a more generic name like `alloc_resources`
- `alloc_resources(...)` will be a class method taking keyword arguments to configure things, and the possible arguments depend on the MCU series (eg F4/F7 just selects CAN2 start bank like, on H7 there is more to configure)
- calling `alloc_resources()` is optional; if it's not called then there are sensible defaults, eg CAN(1) uses all/half available resources

#### Option 2
- remove `initfilterbanks` completely
- add new keyword args to the CAN constructor to specify how many resources it should use, eg `num_filter_banks`
- specifying these args is optional and they default to something sensible, eg all/half resources

----------

Option 2 seems more in line with the rest of the machine API, eg specifying UART buffer size (which comes from the heap, but conceptually is the same as allocating general resources).

iabdalkader committed:
stm32/pyb_can.c: Remove initfilterbanks.

* Add keyword arg to configure the filter bank split.

dpgeorge commented:
> I went with option 2 and removed `initfilterbanks` completely, added a `num_filter_banks` keyword arg for classic F4/F7 CAN, which provides the same functionality, and updated the docs accordingly

Fantastic!

> For FDCAN the configuration remains static. I think the default config is good for now

Yes I agree, let's leave it static until someone requests for it to be configurable.

iabdalkader commented:
Not sure why CI is failing, let me know if I need to fix something.

dpgeorge commented:
There's so much CI, something is always going wrong... but in this case it is #8301 so nothing to worry about for this PR.

dpgeorge commented:
Rebased and merged in ff287d085f95ab0ca3fdf67ee9d8cce6f046b01e through 5cdf9645711cc12b45e602ef5795c33895116fc4.  And I updated the tests in 71344c15f4ef6efc62fbaeef830fa357ee30dce0

Thank you @iabdalkader for your efforts on this, I think it is a good improvement to the API and good to have better FD CAN support!
